src/data/buffer/parser/numeral.lean

```diff
@@ -0,0 +1,112 @@
+/-
+Copyright (c) 2020 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import data.fintype.card
+
+/-!
+# Numeral parsers
+
+This file expands on the existing `nat : parser ℕ` to provide parsers into any type `α` that
+can be represented by a numeral, which relies on `α` having a 0, 1, and addition operation.
+There are also convenience parsers that ensure that the numeral parsed in is not larger than
+the cardinality of the type `α` , if it is known that `fintype α`. Additionally, there are
+convenience parsers that parse in starting from "1", which can be useful for positive ordinals;
+or parser from a given character or character range.
+
+## Main definitions
+
+* 'numeral` : The parser which uses `nat.cast` to map the result of `parser.nat` to the desired `α`
+* `numeral.of_fintype` :  The parser which `guard`s to make sure the parsed numeral is within the
+  cardinality of the target `fintype` type `α`.
+
+## Implementation details
+
+When the `numeral` or related parsers are invoked, the desired type is provided explicitly. In many
+cases, it can be inferred, so one can write, for example
+```lean
+def get_fin : string → fin 5 :=
+sum.elim (λ _, 0) id ∘ parser.run_string (parser.numeral.of_fintype _)
+```
+
+In the definitions of the parsers (except for `numeral`), there is an explicit `nat.bin_cast`
+instead an explicit or implicit `nat.cast`
+-/
+
+open parser parse_result
+
+namespace parser
+
+variables (α : Type) [has_zero α] [has_one α] [has_add α]
+
+/--
+Parse a string of digits as a numeral while casting it to target type `α`.
+-/
+def numeral : parser α :=
+nat.bin_cast <$> nat
+
+/--
+Parse a string of digits as a numeral while casting it to target type `α`,
+which has a `[fintype α]` constraint. The parser ensures that the numeral parsed in
+is within the cardinality of the type `α`.
+-/
+def numeral.of_fintype [fintype α] : parser α :=
+do
+  c ← nat,
+  decorate_error ("<numeral less than " ++ to_string (fintype.card α) ++ ">")
```

Comment:
You can use `sformat!` to make these messages easier to write and read, here and below:
```suggestion
  decorate_error (sformat!"<numeral less than {to_string (fintype.card α)}>")
```